File: select-sv-results/c++/Graphics.cpp

#include "Graphics.h" 

#include <vtkCellPicker.h>
//
#include <vtkExtractSelectedPolyDataIds.h>
#include <vtkExtractSelection.h>
//
#include <vtkGenericCell.h>
//
#include <vtkNamedColors.h>
//
#include <vtkSelection.h>
#include <vtkSelectionNode.h>
//
#include <vtkUnstructuredGrid.h>

//----------
// Graphics
//----------
//
Graphics::Graphics()
{
  // Add a renderer.
  m_Renderer = vtkSmartPointer<vtkRenderer>::New();
  m_Renderer->SetBackground(1.0, 1.0, 1.0);

  // Add a render window.
  m_RenderWindow = vtkSmartPointer<vtkRenderWindow>::New();
  m_RenderWindow->AddRenderer(m_Renderer);
  m_RenderWindow->SetSize(1000, 1000);

  // Add window interactor to use trackball and intercept key presses.
  //
  m_RenderWindowInteractor = vtkSmartPointer<vtkRenderWindowInteractor>::New();
  m_RenderWindowInteractor->SetRenderWindow(m_RenderWindow);

  m_InteractionStyle = vtkSmartPointer<MouseInteractorStyle>::New();
  m_InteractionStyle->SetGraphics(this);
  m_RenderWindowInteractor->SetInteractorStyle(m_InteractionStyle);
  m_InteractionStyle->SetDefaultRenderer(m_Renderer);
  m_RenderWindowInteractor->SetInteractorStyle(m_InteractionStyle);

}

Mesh* Graphics::GetMesh()
{
  return m_Mesh;
}
void Graphics::SetMesh(Mesh* mesh)
{
  m_Mesh = mesh;
}
 

void Graphics::Start()
{
  //m_RenderWindowInteractor->Initialize();
  m_RenderWindowInteractor->Start();
}

//----------------
// CreateGeometry 
//----------------
//
vtkSmartPointer<vtkActor> Graphics::CreateGeometry(vtkSmartPointer<vtkPolyData> polyData)
{
  vtkSmartPointer<vtkDataSetMapper> mapper = vtkSmartPointer<vtkDataSetMapper>::New();
  mapper->SetInputData(polyData);
  mapper->ScalarVisibilityOff();
  vtkSmartPointer<vtkActor> actor = vtkSmartPointer<vtkActor>::New();
  actor->SetMapper(mapper);
  return actor;
}

//-------------
// AddGeometry
//-------------
//
void Graphics::AddGeometry(vtkSmartPointer<vtkActor> geom)
{
  m_Renderer->AddActor(geom);
}

std::string Graphics::GetDataName() 
{
  return m_DataName;
}

void Graphics::SetDataName(std::string name) 
{
  m_DataName = name;
}


/////////////////////////////////////////////////////////////////
//            M o u s e I n t e r a c t o r S t y l e          //
/////////////////////////////////////////////////////////////////

MouseInteractorStyle::MouseInteractorStyle()
{
  m_SelectedMapper = vtkSmartPointer<vtkDataSetMapper>::New();
  m_SelectedActor = vtkSmartPointer<vtkActor>::New();
  m_Colors = vtkSmartPointer<vtkNamedColors>::New();

}

void MouseInteractorStyle::SetGraphics(Graphics* graphics)
{
  m_Graphics = graphics;
}

// Process a keyboard press event.
//
void MouseInteractorStyle::OnKeyPress() 
{
  // Get the keypress.
  vtkRenderWindowInteractor *rwi = this->Interactor;
  std::string key = rwi->GetKeySym();
  std::cout << "Pressed key: " << key << std::endl;

  if ((key == "Escape") || (key == "q")) {
    exit(0);
  }
  else if (key == "s") {
    SelectCell();
  }

  // Forward events
  //vtkInteractorStyleTrackballCamera::OnLeftButtonDown();

}

//------------
// SelectCell
//------------
//
void MouseInteractorStyle::SelectCell() 
{
  std::cout << "---------- Select Mesh Cell ----------" << std::endl;
  vtkSmartPointer<vtkNamedColors> colors = vtkSmartPointer<vtkNamedColors>::New();

  // Get the location of the click (in window coordinates)
  int* pos = this->GetInteractor()->GetEventPosition();

  vtkSmartPointer<vtkCellPicker> picker = vtkSmartPointer<vtkCellPicker>::New();
  picker->SetTolerance(0.0005);

  // Pick from this location.
  picker->Pick(pos[0], pos[1], 0, this->GetDefaultRenderer());

  double* worldPosition = picker->GetPickPosition();
  auto cellID = picker->GetCellId();

  if (cellID == -1) {
    return;
  }

  std::cout << "Pick position is: " << worldPosition[0] << " " << worldPosition[1] << " " << worldPosition[2] << endl;
  SelectMesh(cellID);
}

//--------------
// AddSelection
//--------------
//
void MouseInteractorStyle::SelectMesh(int cellID)
{ 
  vtkSmartPointer<vtkIdTypeArray> ids = vtkSmartPointer<vtkIdTypeArray>::New();
  ids->SetNumberOfComponents(1);
  ids->InsertNextValue(cellID);

  vtkSmartPointer<vtkSelectionNode> selectionNode = vtkSmartPointer<vtkSelectionNode>::New();
  selectionNode->SetFieldType(vtkSelectionNode::CELL);
  selectionNode->SetContentType(vtkSelectionNode::INDICES);
  selectionNode->SetSelectionList(ids);

  vtkSmartPointer<vtkSelection> selection = vtkSmartPointer<vtkSelection>::New();
  selection->AddNode(selectionNode);

  if (m_Graphics->GetMesh()->IsSurface()) {
    SelectSurfaceMesh(cellID, selection);
  }

}

//-------------------
// SelectSurfaceMesh
//-------------------
//
void MouseInteractorStyle::SelectSurfaceMesh(int cellID, vtkSmartPointer<vtkSelection> selection)
{ 
  auto mesh = m_Graphics->GetMesh();
  vtkPolyData* polydata = vtkPolyData::SafeDownCast(mesh->GetMesh());
  auto dataName = m_Graphics->GetDataName();
  auto data = mesh->GetDataArray(dataName);

  // Extract a list of cells from polydata.
  //
  vtkSmartPointer<vtkExtractSelection> extractSelection = vtkSmartPointer<vtkExtractSelection>::New();
  //vtkSmartPointer<vtkExtractSelectedPolyDataIds> extractSelection = vtkSmartPointer<vtkExtractSelectedPolyDataIds>::New();
  extractSelection->SetInputData(0, polydata); 
  extractSelection->SetInputData(1, selection);
  extractSelection->Update();

  // Get selected cells as a vtkPolyData.
  //
  vtkSmartPointer<vtkUnstructuredGrid> selected = vtkSmartPointer<vtkUnstructuredGrid>::New();
  //vtkSmartPointer<vtkPolyData> selected = vtkSmartPointer<vtkPolyData>::New();
  selected->ShallowCopy(extractSelection->GetOutput());
  std::cout << "There are " << selected->GetNumberOfPoints() << " points in the selection." << std::endl;
  std::cout << "There are " << selected->GetNumberOfCells() << " cells in the selection." << std::endl;

   // Print cell ID and connectivity.
   //
   vtkGenericCell* cell = vtkGenericCell::New();
   polydata->GetCell(cellID, cell);
   auto numPts = cell->GetNumberOfPoints();
   std::cout << "Cell ID is: " << cellID << std::endl;
   std::cout << "  Number of cell points: " << numPts << std::endl;
   std::cout << "  Connectivity: ";
   for (vtkIdType pointInd = 0; pointInd < numPts; ++pointInd) {
     auto id = cell->PointIds->GetId(pointInd);
     std::cout << id << " ";
   }
   std::cout << std::endl;

   // Print cell data.
   std::cout << std::endl;
   std::cout << "Cell Data: " << std::endl;
   std::cout << "  Name: " << dataName << std::endl;
   for (vtkIdType pointInd = 0; pointInd < numPts; ++pointInd) {
     auto id = cell->PointIds->GetId(pointInd);
     double value = data->GetValue(id);
     std::cout << "  ID: " << id << "   value: " << value << std::endl;
   }

  // Display the selected cell.
  //
  m_SelectedMapper->SetInputData(selected);
  m_SelectedMapper->ScalarVisibilityOff();
  m_SelectedActor->SetMapper(m_SelectedMapper);
  //selectedActor->GetProperty()->EdgeVisibilityOn();
  //m_SelectedActor->GetProperty()->SetRepresentationToWireframe();
  m_SelectedActor->GetProperty()->SetColor(m_Colors->GetColor3d("Green").GetData());
  m_SelectedActor->GetProperty()->SetLineWidth(4);

  this->Interactor->GetRenderWindow()->GetRenderers()->GetFirstRenderer()->AddActor(m_SelectedActor);
  this->Interactor->GetRenderWindow()->Render();
}


//void MouseInteractorStyle::OnLeftButtonDown()
//{
//}

vtkStandardNewMacro(MouseInteractorStyle);

File: create-surface-segmentations/python/graphics.py
import sv
import vtk
from math import sqrt

class MouseInteractorStyle(vtk.vtkInteractorStyleTrackballCamera):

    def __init__(self, surface=None, picking_keys=None, event_table=None):
        #self.AddObserver("LeftButtonPressEvent", self.leftButtonPressEvent)
        self.AddObserver("KeyPressEvent", self.onKeyPressEvent)
        self.AddObserver("CharEvent", self.onCharEvent)
        self.window = None
        self.renderer = None
        self.surface = surface 
        self.event_table = event_table
        self.selected_points = []
        self.selected_node_ids = []
        self.selected_node_id = None
        self.selected_cell_id = None
        self.picking_keys = picking_keys 
        self.picked_actor = None
        self.last_picked_actor = None
        self.method_queue = []
        self.pick_glyph_actor_queue = []

    def select_event(self, renderer, key, obj, event):
        '''Process a select event. 
        '''
        clickPos = self.GetInteractor().GetEventPosition()
        picker = vtk.vtkCellPicker()
        picker.Pick(clickPos[0], clickPos[1], 0, self.renderer)

        # Get selecected geometry. 
        self.picked_actor = picker.GetActor()
        if self.picked_actor == None:
            #self.OnLeftButtonDown()
            return
        position = picker.GetPickPosition()
        cell_id = picker.GetCellId()

        if cell_id == -1: 
            return

        print(" ")
        print("Picked position: {0:g} {1:g} {2:g} ".format(position[0], position[1], position[2]))
        print("Cell id is: {0:d}".format(cell_id))

        min_i = -1
        min_d = 1e5
        min_p = 3*[0.0]
        surface = self.surface
        points = surface.GetPoints()

        for i in range(points.GetNumberOfPoints()):
            p = 3*[0.0]
            points.GetPoint(i,p)
            dx = p[0] - position[0]
            dy = p[1] - position[1]
            dz = p[2] - position[2]
            d = sqrt(dx*dx + dy*dy + dz*dz)
            if d < min_d:
                min_d = d
                min_p[0] = p[0]
                min_p[1] = p[1]
                min_p[2] = p[2]
                min_i = i  

        print("Picked node: {0:d} {1:g} {2:g} {3:g} ".format(min_i, min_p[0], min_p[1], min_p[2]))
        self.selected_node_ids.append(min_i)
        self.selected_node_id = min_i
        self.selected_cell_id = cell_id

        ## Show picked point.
        points = vtk.vtkPoints()
        points.SetNumberOfPoints(1)
        points.SetPoint(0, position[0], position[1], position[2])
        geom = vtk.vtkPolyData()
        geom.SetPoints(points)
        glyphFilter = vtk.vtkVertexGlyphFilter()
        glyphFilter.SetInputData(geom)
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputConnection(glyphFilter.GetOutputPort())
        actor = vtk.vtkActor()
        if key == 's':
            actor.GetProperty().SetColor(0.0, 1.0, 0.0)
        elif key == 't':
            actor.GetProperty().SetColor(1.0, 0.0, 0.0)
        else:
            actor.GetProperty().SetColor(1.0, 1.0, 1.0)
        actor.GetProperty().SetPointSize(10)
        actor.SetMapper(mapper)
        self.renderer.AddActor(actor)
        self.pick_glyph_actor_queue.append(actor)

        # Get picked face ID.
        data_name = "ModelFaceID" 
        cell_data = surface.GetCellData().GetArray(data_name)
        if cell_data != None:
            face_id = cell_data.GetValue(cell_id)
            print("Picked face: {0:d} ".format(face_id))

        #self.window.Render()

    def onKeyPressEvent(self, renderer, event):
        '''Process a key press event.
        '''
        key = self.GetInteractor().GetKeySym()

        if key == 'u' and len(self.method_queue) != 0:
            last_method = self.method_queue.pop()
            last_method(undo=True) 
            last_actor = self.pick_glyph_actor_queue.pop()
            last_actor.SetVisibility(False)

        if (key in self.picking_keys):
            self.select_event(renderer, key, None, event)

        if self.event_table == None:
            return

        if (key in self.event_table):
            value = self.event_table[key]
            if type(value) == tuple:
                method = value[0]
                data = value[1]
            else:
                method = value
                data = None

            # Call the method with selected node/cell and any data.
            #print("###### self.selected_node_id: " + str(self.selected_node_id))
            method(surface=self.surface, node_id=self.selected_node_id, cell_id=self.selected_cell_id, data=data, renderer=self.renderer)

            # Store the last method so we can use it for an undo operation.
            self.method_queue.append(method)

        self.window.Render()

    def onCharEvent(self, renderer, event):
        '''Process an on char event.

        This is used to prevent passing the shortcut key 'w' to vtk which we use
        to write selected results and vtk uses to switch to wireframe display. 
        '''
        key = self.GetInteractor().GetKeySym()
        if ((key != 'w') and (key != 'e')):
            self.OnChar()

def add_line(renderer, pt1, pt2, color=[1.0, 1.0, 1.0], width=2):
    line = vtk.vtkLineSource()
    line.SetPoint1(pt1);
    line.SetPoint2(pt2)
    line.Update()
    polydata = line.GetOutput()
    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputData(polydata)
    mapper.ScalarVisibilityOff()
    actor = vtk.vtkActor()
    actor.SetMapper(mapper)
    actor.GetProperty().SetColor(color[0], color[1], color[2])
    actor.GetProperty().SetLineWidth(width)
    renderer.AddActor(actor)

def display(renderer_win):
    interactor = vtk.vtkRenderWindowInteractor()
    interactor.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())
    interactor.SetRenderWindow(renderer_win)
    # Set the window title.
    renderer_win.Render()
    renderer_win.SetWindowName("SV Python API")
    interactor.Start()

def add_geometry(renderer, polydata, color=[1.0, 1.0, 1.0], line_width=1, wire=False, edges=False, array_name=None, scalar_range=None):
    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputData(polydata)
    if array_name == None:
        mapper.SetScalarVisibility(False)
    else:  
        mapper.SetArrayName(array_name)
        mapper.SelectColorArray(array_name)
        mapper.SetScalarVisibility(True)
        mapper.SetScalarModeToUseCellFieldData()
        mapper.SetScalarRange(scalar_range)
    actor = vtk.vtkActor()
    actor.SetMapper(mapper)
    actor.GetProperty().SetColor(color[0], color[1], color[2])
    #actor.GetProperty().SetPointSize(5)
    #actor.GetProperty().BackfaceCullingOn();

    if wire:
        actor.GetProperty().SetRepresentationToWireframe()
        actor.GetProperty().SetLineWidth(line_width)
    elif not edges:
        actor.GetProperty().SetLineWidth(line_width)

    if edges:
        actor.GetProperty().EdgeVisibilityOn();
    else:
        actor.GetProperty().EdgeVisibilityOff();

    renderer.AddActor(actor)
    return actor

def add_glyph_points(renderer, points, color=[1.0, 1.0, 1.0], size=2):
    geom = vtk.vtkPolyData()
    geom.SetPoints(points)
    glyphFilter = vtk.vtkVertexGlyphFilter()
    glyphFilter.SetInputData(geom)
    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputConnection(glyphFilter.GetOutputPort())
    actor = vtk.vtkActor()
    actor.GetProperty().SetColor(color)
    actor.GetProperty().SetPointSize(size)
    actor.SetMapper(mapper)
    renderer.AddActor(actor)

def add_sphere(renderer, center, radius, color=[1.0, 1.0, 1.0], wire=False):
    sphere = vtk.vtkSphereSource()
    sphere.SetCenter(center[0], center[1], center[2])
    sphere.SetRadius(radius) 
    sphere.SetPhiResolution(16)
    sphere.SetThetaResolution(16)
    sphere.Update()
    sphere_pd = sphere.GetOutput() 
    return add_geometry(renderer, sphere_pd, color, wire=wire)

def init_graphics(win_width, win_height):
    ''' Create renderer and graphics window.
    '''
    renderer = vtk.vtkRenderer()
    renderer_win = vtk.vtkRenderWindow()
    renderer_win.AddRenderer(renderer)
    renderer.SetBackground(0.8, 0.8, 0.8)
    renderer_win.SetSize(win_width, win_height)
    #renderer_win.Render()
    #renderer_win.SetWindowName("SV Python API")
    print("Create renderer and graphics window.")
    #print("---------- Alphanumeric Keys ----------")
    print("q - Quit")
    #print("s - Select a face.")
    return renderer, renderer_win 

def init_picking(window, renderer, surface, picking_keys, event_table=None):

    # Create a trackball interacter to transoform the geometry using the mouse.
    interactor = vtk.vtkRenderWindowInteractor()
    interactor.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())
    interactor.SetRenderWindow(window)

    face_ids = surface.GetCellData().GetArray("ModelFaceID")
    if face_ids == None:
        print("No ModelFaceID data.")
    else:
        face_ids_range = 2*[0]
        face_ids.GetRange(face_ids_range, 0)
        min_id = int(face_ids_range[0])
        max_id = int(face_ids_range[1])
        print("Face IDs range: {0:d} {1:d}".format(min_id, max_id))

    # Add the custom style.
    style = MouseInteractorStyle(surface, picking_keys, event_table)
    style.window = window
    style.renderer = renderer
    interactor.SetInteractorStyle(style)
    style.SetCurrentRenderer(renderer)
    return interactor


